### DISCUSSION TITLE
ESPHome OpenTherm master with a mix of underfloor and radiators

### DISCUSSION THREAD

**Discussion by clau-bucur** [Alexwijn/SAT#55] [General]
I've just started my OpenTherm journey and discovered this great component. Thank you for it!

### My Setup:
- **Boiler**: Viessmann Vitodens 100-W (it's on the SAT boiler compatibility list)
- **Controller**: OpenTherm master shield running ESPHome with [olegtarasov's component](https://github.com/olegtarasov/esphome-opentherm-component)  
  - Running SAT development version with [cnrd's recommended configuration](https://github.com/Alexwijn/SAT/discussions/50#discussioncomment-10944949)

#### Downstairs:
- One radiator directly connected to the boiler (no valve)
- Underfloor heating for the entire space
  - Single heating zone
  - Temperature-controlled circulating pump
  - One valve (3-minute opening time, on/off only)
  - External temperature sensor
  - `generic_thermostat` climate

The pump and valve are controlled together by a smart relay. When downstairs heating is not needed, the valve closes to prevent hot water circulation through this circuit.

#### Upstairs:
- **Master Valve**: Controls the upstairs circuit, opening only when heating is required in any upstairs room
- **Radiators**:
  - One radiator directly connected to the upstairs circuit (no valve)
  - Three rooms, each with:
    - Radiator + smart TRV + external temperature sensor

I'm replacing current TRVs (basic Chinese ZigBee ones with limited flexibility) with Danfoss Ally, as these support external sensors and offer many more options.

Currently, I've created a `generic_thermostat` for each ZigBee thermostat to use the external temperature sensor. This generic thermostat fully opens or closes the ZigBee thermostat based on heating needs.

So, I have **two independent zones and four independent climate devices**. Each climate has different heating requirements, and I’ve set up schedules for them using [nielsfaber’s scheduler_component](https://github.com/nielsfaber/scheduler-component).

---

In the past (I just started using OpenTherm this heating season), I had an on/off boiler that I manually turned on and off whenever heating was needed. This worked well, though it wasn’t very comfortable, especially upstairs.

Now that I've switched to OpenTherm, I'm finding it hard to understand how to make the most of it. This is when I found `SAT`, and I thought, "This is it!"

So, my plan was to add a new SAT integration entry for each of my four climates, expecting SAT to take over from there. However, things are never so simple. 😄

---

### My Configuration Process

1. **Office Integration**  
   - Created a new integration entry named "Office"
   - Selected the ESPHome device
   - Configured it with office temperature, exterior temperature, and humidity
   - Set the office climate as the primary climate on the Areas page (no room climates)

   As a beginner, it’s very challenging to understand how to configure this page correctly. I would suggest updating the docs with more beginner-friendly guidance, like explaining "The primary entity’s role is this and that; room entities’ roles are this and that." 

   The current docs seem to be written for those already experienced with SAT, which makes it difficult for newcomers like me (it feels like it was written by the author, for the author 😄). Also, the information on the Areas configuration page differs from what’s explained in the docs.

2. **Bedroom Integration**  
   - Tried to create a new integration entry named "Bedroom" and selected the ESPHome device
   - Got the error **"Gateway is already configured"** preventing reuse of the same device for multiple entries

---

### Questions

1. Have I misunderstood how SAT works and should be used?
2. Will this behavior change once ESPHome is fully supported, allowing the same ESPHome device for multiple integration entries?
3. SAT creates a climate entity for each integration entry. How does this work, and how should it be used? (This would be helpful to explain in the docs.) I don’t understand when it switches to the `heating` hvac_action.
4. Given my setup with four independent climates + external temperature sensors, how can a single "inside sensor entity" handle all of them?

Thank you very much for your patience in reading all this!

**Comment by sergeantd83** [comment]
> So, my plan was to add a new SAT integration entry for each of my four climates, expecting SAT to take over from there. However, things are never so simple.

No, these generic thermostats should be in the multiroom tab rooms section during initial configuration.

> The current docs seem to be written for those already experienced with SAT, which makes it difficult for newcomers like me (it feels like it was written by the author, for the author 😄)

Yes, you are correct. We are trying to improve the documentation, soon it will be available. You can have an early sneak peek [here](https://github.com/sergeantd83/SAT). We will appreciate any feedback or recommendations.

> Have I misunderstood how SAT works and should be used?

Maybe the documentation I provided will clear things out.

> Will this behavior change once ESPHome is fully supported, allowing the same ESPHome device for multiple integration entries?

No, you can’t have multiple SAT instances with one boiler.

> SAT creates a climate entity for each integration entry. How does this work, and how should it be used? (This would be helpful to explain in the docs.) I don’t understand when it switches to the heating hvac_action.

You don’t need to create new SAT entities. Just use the already configured generic thermostats.

> Given my setup with four independent climates + external temperature sensors, how can a single "inside sensor entity" handle all of them?

Use as inside temperature sensor the temperature sensor that is placed in your living room.

Some more recommendations:

1) If the overshoot protection value mechanism failed, manually enter the value 50 during the config flow

2) Use these values as starting point
Precision Curve
Improved controller
Heating curve value 1.8
Automatic gains 1.8
Derivative Time Weight 2.1

~~3) Try the latest `develop` version.~~

**Reply by clau-bucur** [reply]
> > So, my plan was to add a new SAT integration entry for each of my four climates, expecting SAT to take over from there. However, things are never so simple.
> 
> No, these generic thermostats should be in the multiroom tab rooms section during initial configuration.

Alright, I've configured it like this, **is it correct?**
![image](https://github.com/user-attachments/assets/4357e688-1da3-4180-a864-042d15a261d5)
As per my understanding from the docs you linked below, the Primary entity in this page refers to a physical OpenTherm thermostat which is directly connected to the boiler (if that exists in one's setup and is integrated in Home Assistant).
**Am I correct?**

> > The current docs seem to be written for those already experienced with SAT, which makes it difficult for newcomers like me (it feels like it was written by the author, for the author 😄)
> 
> Yes, you are correct. We are trying to improve the documentation, soon it will be available. You can have an early sneak peek [here](https://github.com/sergeantd83/SAT). We will appreciate any feedback or recommendations.

Some suggestions:
  - **Configuration steps**: I believe it will be easier for new users to understand the configuration process if the docs steps also refer to each configuration page by it's name.
    Eg. You have in the docs, section Configuration / OpenTherm, step "4. Multi-room setup". It should be clear from the beginning that it refers to the Areas configuration flow page.
  - **Heating System configuration**: Cover the case for mixed systems, like for example my situation where I have a combination of Radiators and Underfloor.
  - **Areas configuration**: The explanation for the "Primary" entity, IMHO, could use improving. I would use explaining how is the SAT operation affected/changed if the Primary is configured or not.
  - **Operational information**: Once configured SAT will create a climate entity.
      Would be nice to understand and explained in the docs:
        - How is SAT working when changing the setpoint temperature for this climate?
          After configured SAT climate has a 7* setpoint.
          I see the climate switches to "heating" state based on this temperature setpoint, even though my rooms climates are already in a "heating" climate.
          It switches to "heating" later if setpoint is low, and sooner if setpoint is higher.
          What is the algorithm controlling this behaviour?

> > SAT creates a climate entity for each integration entry. How does this work, and how should it be used? (This would be helpful to explain in the docs.) I don’t understand when it switches to the heating hvac_action.
> 
> You don’t need to create new SAT entities. Just use the already configured generic thermostats.

When I switch to using Danfoss TRVs I will drop the generic thermostats and use Danfoss directly.
**Will this impact SAT's behaviour in any way?**

Important question: **SAT does not attempt to change the room climate entities** state in any way, right? Like changing the hvac_mode, hvac_action or adjust the valve position.
SAT is only reading the climate state and act accordingly.

> > Given my setup with four independent climates + external temperature sensors, how can a single "inside sensor entity" handle all of them?
> 
> Use as inside temperature sensor the temperature sensor that is placed in your living room.

An important question here: What type of heating system would be optimal that I choose in my case: **Radiator or Underfloor?**
How would SAT behave differently in my case if I select one or the other? (nice to be included in the docs)

> 
> Some more recommendations:
> 
>     1. If the overshoot protection value mechanism failed, manually enter the value 50 during the config flow
> 
>     2. Use these values as starting point
>        Precision Curve
>        Improved controller
>        Heating curve value 1.8

I suspect you refer to "Heating Curve Coefficient" from the Configure / General page.

>        Automatic gains 1.8
>        Derivative Time Weight 2.1

These two settings only accept whole numbers so I can't set the values you suggested.

**Comment by sergeantd83** [comment]
Sorry for my late response and thank you for the suggestions!

>As per my understanding from the docs you linked below, the Primary entity in this page refers to a physical OpenTherm thermostat which is directly connected to the boiler (if that exists in one's setup and is integrated in Home Assistant).
Am I correct?

Yes you are correct.

>As per my understanding from the docs you linked below, the Primary entity in this page refers to a physical OpenTherm thermostat which is directly connected to the boiler (if that exists in one's setup and is integrated in Home Assistant).
Am I correct?

It refers to ANY physical thermostat ( not only OpenTherm ) that is directly connected to the boiler and integrated in Home Assistant.

>When I switch to using Danfoss TRVs I will drop the generic thermostats and use Danfoss directly.
Will this impact SAT's behaviour in any way?

Yes, you will add directly the Danfoss TRVs and SAT will monitor the Danfoss generated climate entity.

>Important question: SAT does not attempt to change the room climate entities state in any way, right? Like changing the hvac_mode, hvac_action or adjust the valve position.
SAT is only reading the climate state and act accordingly.

Exactly, we don’t interfere with the control of the TRVs at any point. We only monitor the `hvac_action` and the `error` ( Room Setpoint - Room Temperature ).

>An important question here: What type of heating system would be optimal that I choose in my case: Radiator or Underfloor?
How would SAT behave differently in my case if I select one or the other? (nice to be included in the docs)

I think you should try the Radiators option and if overshoots change to underfloor. You can see the behavior of both calculations in the [heating curve hyperlinks](https://github.com/sergeantd83/SAT?tab=readme-ov-file#general-tab).

>>   Automatic gains 1.8
>> Derivative Time Weight 2.1
>
>These two settings only accept whole numbers so I can't set the values you suggested.

Enter these values manually, it will be fixed in the next version.

**Reply by clau-bucur** [reply]
Thank you for your answers.
I've entered the configuration values manually and let's see now how it works in the next days :)

One last thing is left unclear to me at the moment, if it's possible to explain:
How is SAT working when changing the setpoint temperature for it's climate entity?
After configured SAT climate has a 7 °C target temperature.
I see the climate switches to `heating` state differently, based on this temperature target, even though my rooms climates are already in a `heating` climate.
It switches to `heating` later if the target temperature is low, and sooner if the target is higher.

**Reply by sergeantd83** [reply]
> How is SAT working when changing the setpoint temperature for it's climate entity?

You need to sync your Living Room ( Downstairs ) with the SAT climate entity. So when you change the SAT climate entity the Living Room thermostat will follow the change. You can do it also with the` Synchronize climates with preset (sleep / away / activity)` option in the `Presets` Tab.

> After configured SAT climate has a 7 °C target temperature.

You can change it and choose the desired setpoint.

> I see the climate switches to heating state differently, based on this temperature target, even though my rooms climates are already in a heating climate.

This is the Low Load control ( probably ) operation. You can look at developer tools  -> State -> SAT climate entity. All the values you need are there as attributes.

> It switches to heating later if the target temperature is low, and sooner if the target is higher.

I'm sorry I can't understand. Can you please provide some graphs? Please include in the graph the values of the below sensors:
- Room Setpoint
- Room Temperature
- Boiler Targer Temperature
- Boiler Flow Water Temperature
- Max modulation
- Relative Modulation

**Reply by clau-bucur** [reply]
> > It switches to heating later if the target temperature is low, and sooner if the target is higher.
> 
> I'm sorry I can't understand. Can you please provide some graphs? Please include in the graph the values of the below sensors:
> 
>     * Room Setpoint
> 
>     * Room Temperature
> 
>     * Boiler Targer Temperature
> 
>     * Boiler Flow Water Temperature
> 
>     * Max modulation
> 
>     * Relative Modulation

I can't seem to be able to reproduce this anymore.

One more question, I hope the last for this discussion:
It's mentioned that the automatically calculation of the OPV value should take about 20 minutes.
For me it takes about 5 minutes. I've done that more than a couple of times, always the same duration.
However the OPV is never the same. Sometimes around 55, sometimes 28 and once even went to 62 (higher than the Maximum Setpoint of the boiler which was set to 60).
Is this normal?

**Reply by sergeantd83** [reply]
> I can't seem to be able to reproduce this anymore.

I may can explain this behavior. When the room temperature is outside of the deadband ( ± 0.1 ) SAT disables the integral part. Probably this caused this issue.

> It's mentioned that the automatically calculation of the OPV value should take about 20 minutes.
For me it takes about 5 minutes. I've done that more than a couple of times, always the same duration.
However the OPV is never the same. Sometimes around 55, sometimes 28 and once even went to 62 (higher than the Maximum Setpoint of the boiler which was set to 60).
Is this normal?

You are correct this is not normal at all. We need to fix the overshoot protection mechanism. You can use the manual method in this [discussion](https://github.com/Alexwijn/SAT/discussions/36) at the moment. As Maximum temperature use the value 60. We will fix it soon.
